File: source/c/arbol.c
// ==========================================
// ESTRUCTURA DE DATOS: ÁRBOL BINARIO DE BÚSQUEDA
// ==========================================
// Explicación Detallada:
// En C, usamos punteros para enlazar los nodos, parecido a la lista enlazada,
// pero con dos ramas. El nodo raíz es nuestro punto de entrada al árbol.
// La regla es: hijos izquierdos < padre < hijos derechos.

#include <stdio.h>
#include <stdlib.h>

// Definición de Nodo
struct Nodo {
    int dato;
    struct Nodo *izq, *der;
};

// CREAR NODO
// Función auxiliar que pide memoria y setea valores por defecto para los hijos (NULL).
struct Nodo* crearNodo(int dato) {
    struct Nodo* nuevo = (struct Nodo*)malloc(sizeof(struct Nodo));
    nuevo->dato = dato;
    nuevo->izq = nuevo->der = NULL;
    return nuevo;
}

// INSERCIÓN
// Proceso (Recursivo):
// Recibe un nodo raíz local. Si está vacío, ahí inserta el nuevo nodo y lo devuelve.
// Si el dato es menor, llama recursivamente a 'insertar' en la rama izquierda.
// Si es mayor, lo hace en la rama derecha.
struct Nodo* insertar(struct Nodo* nodo, int dato) {
    if (nodo == NULL) return crearNodo(dato); // Espacio libre encontrado
    
    if (dato < nodo->dato)
        nodo->izq = insertar(nodo->izq, dato); // Bajar por izquierda
    else if (dato > nodo->dato)
        nodo->der = insertar(nodo->der, dato); // Bajar por derecha
        
    return nodo;
}

// RECORRIDO (Inorden)
// Proceso: 
// 1. Recorrer rama izquierda
// 2. Visitar el nodo
// 3. Recorrer rama derecha
// Esto garantiza imprimir los valores ordenados ascendentemente.
void inorden(struct Nodo* raiz) {
    if (raiz != NULL) {
        inorden(raiz->izq);
        printf("%d ", raiz->dato);
        inorden(raiz->der);
    }
}

// BÚSQUEDA
// Proceso recursivo similar a la inserción, navegando según si el dato 
// es menor (izquierda) o mayor (derecha) al valor del nodo actual.
int buscar(struct Nodo* raiz, int dato) {
    if (raiz == NULL) return 0; // Falso, llegó al final y no lo encontró
    if (raiz->dato == dato) return 1; // Verdadero, hallado
    
    if (dato < raiz->dato)
        return buscar(raiz->izq, dato); // Buscar a la izquierda
    else
        return buscar(raiz->der, dato); // Buscar a la derecha
}

int main() {
    struct Nodo* raiz = NULL;
    // La primera inserción inicializa la raíz
    raiz = insertar(raiz, 50);
    insertar(raiz, 30);
    insertar(raiz, 70);
    insertar(raiz, 20);
    insertar(raiz, 40);
    
    printf("Recorrido inorden (ordenado): ");
    inorden(raiz);
    
    printf("\n¿El 40 está en el árbol?: %s\n", buscar(raiz, 40) ? "Sí" : "No");
    return 0;
}

File: source/python/arbol.py
# ==========================================
# ESTRUCTURA DE DATOS: ÁRBOL BINARIO DE BÚSQUEDA (BST)
# ==========================================
# Explicación Detallada:
# Un árbol es una estructura jerárquica y no lineal. En un Árbol Binario de Búsqueda,
# cada nodo tiene como máximo dos hijos (izquierdo y derecho).
# La regla clave (propiedad de búsqueda) es:
# - Todos los datos en el subárbol izquierdo son MENORES que el dato del nodo raíz.
# - Todos los datos en el subárbol derecho son MAYORES que el dato del nodo raíz.
# Esto hace que la búsqueda de elementos sea sumamente rápida, comparable a una 
# búsqueda binaria.

class NodoArbol:
    # Cada nodo guarda un valor y enlaces a sus posibles dos hijos.
    def __init__(self, dato):
        self.dato = dato
        self.izq = None
        self.der = None

class Arbol:
    def __init__(self):
        # El árbol comienza con una raíz vacía.
        self.raiz = None

    # INSERCIÓN
    # Proceso:
    # 1. Si la raíz está vacía, el nuevo dato forma la raíz.
    # 2. Si no, usamos un método recursivo interno para buscar su lugar.
    #    Si el dato es menor al nodo actual, nos movemos a la izquierda.
    #    Si el dato es mayor, nos movemos a la derecha, hasta encontrar un espacio libre (None).
    def insertar(self, dato):
        if not self.raiz:
            self.raiz = NodoArbol(dato)
        else:
            self._insertar(self.raiz, dato)

    def _insertar(self, nodo, dato):
        if dato < nodo.dato:
            if nodo.izq:
                self._insertar(nodo.izq, dato) # Seguir bajando
            else:
                nodo.izq = NodoArbol(dato)     # Insertar hoja
        else:
            if nodo.der:
                self._insertar(nodo.der, dato) # Seguir bajando
            else:
                nodo.der = NodoArbol(dato)     # Insertar hoja

    # RECORRIDO (Inorden)
    # Proceso: Visita el subárbol izquierdo, luego la raíz actual, luego el derecho.
    # En un árbol binario de búsqueda, ¡el recorrido Inorden imprime los valores ORDENADOS!
    def inorden(self, nodo):
        if nodo:
            self.inorden(nodo.izq)       # 1. Todo lo de la izquierda
            print(nodo.dato, end=" ")    # 2. El nodo en sí
            self.inorden(nodo.der)       # 3. Todo lo de la derecha

    # BÚSQUEDA
    # Proceso recursivo:
    # 1. Si el nodo actual es el buscado (o es None), terminamos.
    # 2. Si es menor, buscamos en el lado izquierdo.
    # 3. Si es mayor, buscamos en el derecho.
    def buscar(self, nodo, dato):
        if not nodo or nodo.dato == dato:
            return nodo is not None
        if dato < nodo.dato:
            return self.buscar(nodo.izq, dato)
        return self.buscar(nodo.der, dato)

if __name__ == "__main__":
    a = Arbol()
    # Insertando valores (50 va primero y es la raíz)
    a.insertar(50)
    a.insertar(30) # Va a la izquierda del 50
    a.insertar(70) # Va a la derecha del 50
    a.insertar(20) # Va a la izquierda del 30
    a.insertar(40) # Va a la derecha del 30

    print("Recorrido Inorden (Los datos deben salir de menor a mayor):")
    a.inorden(a.raiz)
    print("\n\n¿El 40 está en el árbol?:", a.buscar(a.raiz, 40))
    print("¿El 99 está en el árbol?:", a.buscar(a.raiz, 99))